bin/specextract

```diff
@@ -79,381 +81,449 @@ v4 = make_verbose_level(toolname, 4)
+
+            mkwarf.punlearn()
+
+            mkwarf.outfile = arffile
+            mkwarf.weightfile = weightfile.name
+            mkwarf.mskfile = mskfile
+            mkwarf.dafile = dafile
+            mkwarf.spectrumfile = ""
+            mkwarf.egridspec = ebin
+            mkwarf.detsubsysmod = "BPMASK=0x03ffff"
+            mkwarf.clobber = "yes"
+            mkwarf.verbose = str(verbose)
+
+            try:
+                mkwarf.infile = f"{tdetwmap.name}[wmap]"
+            except AttributeError:
+                mkwarf.infile = f"{tdetwmap}[wmap]"
 
-    ## pbkkeys are: 'OCLKPAIR', 'ORC_MODE', 'SUM_2X2', 'FEP_CCD'
-    ## asol-derived keys are: 'DY_AVG', 'DZ_AVG', 'DTH_AVG'
+            mkwarf()
+			
+    except Exception as err_msg:
+        print(err_msg)
 
-    pbk_kw = ("OCLKPAIR","ORC_MODE","SUM_2X2","FEP_CCD")
-    # asp_kw = ["DY_AVG","DZ_AVG","DTH_AVG"]
+        raise IOError(f"Failure to create weighted ARF.  Possible causes include: zero counts in the input region; a memory allocation error; or corrupt ARDLIB.  Try running {toolname} with weight=no, binarfmap!=1, or punlearn ardlib.")
 
-    file_status = []
+    return arffile, weightfile
 
-    for inf in infile_stack:
-        inf = fileio.get_file(inf)
 
-        headerkeys = fileio.get_keys_from_file(inf)
 
-        if headerkeys["INSTRUME"] == "ACIS":
-            try:
-                verify_keys = []
+def correct_arf(full_outroot, infile, evt_filename, orig_arf,
+                skyx, skyy, binarfcorr, clobber):
+    """
+    Apply an energy-dependent point-source aperture correction
+    to the source ARF created by mkarf, if user has set the
+    'correct' specextract parameter to 'yes' and weight is 'no'.
+    It is not appropriate to run arfcorr on background ARFs because
+    background is extended.
 
-                verify_keys.extend([headerkeys.has_key(kw) for kw in pbk_kw])
+    Return the name of the corrected ARF file.
+    """
+    
+    gz_input = os.path.exists(evt_filename+".gz")
+    guz_input = os.path.exists(evt_filename)
 
-                if False in verify_keys:
-                    v1(f"WARNING: {fileio.remove_path(inf)} missing header keywords.\n")
+    if not infile.endswith(".gz") and not guz_input:
 
-                    file_status.append(inf)
+        if gz_input:
+            evt_filename_gz = f"{evt_filename}.gz"
 
-            except AttributeError:
-                pbk_kw_list = list(pbk_kw)
+            infile = infile.replace(evt_filename, evt_filename_gz)
 
-                verify_keys = headerkeys.keys() & pbk_kw_list # find elements in common
+            v1("NOTE: {0} does not exist, but {0}.gz does; using the latter as input to arfcorr for the ARF correction.\n".format(evt_filename))
 
-                for s in verify_keys:
-                    pbk_kw_list.remove(s)
+    # arfcorr required input image:
+    reg_image  = f"{infile}[bin sky={binarfcorr}]"
```

Comment:
A question about this (and this is the same as what was done before as far as I can tell): would it make sense to add an energy filter here so we don't include particle events. Or does it not really matter?

And also is this only done for ACIS data or also HRC (if the latter then it would just complicate how you would filter the event file)?

Reply:
This is also used with PSF-corrected HRC ARFs.  As far as I can tell, arfcorr in ARF-correction mode uses the input image is used to determine the size and scale of the PSF, and the response's energy grid is used for each step of PSF-corrected response.  I'm doubtful that the high-energy particle events would really impact the estimated PSF size.